package Clinic_Management.ClinicalConsultationService.entity;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "encounters", indexes = {
        @Index(name = "idx_encounter_ticket", columnList = "ticket_number"),
        @Index(name = "idx_encounter_patient", columnList = "patient_id"),
        @Index(name = "idx_encounter_doctor", columnList = "doctor_id, encounter_date")
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Encounter {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "ticket_number", nullable = false, length = 20)
    private String ticketNumber;

    @Column(name = "patient_id", nullable = false)
    private Long patientId;

    @Column(name = "patient_name", nullable = false)
    private String patientName;

    @Column(name = "doctor_id", nullable = false)
    private Long doctorId;

    @Column(name = "doctor_name", nullable = false)
    private String doctorName;

    @Column(name = "department_id", nullable = false)
    private Long departmentId;

    @Column(name = "encounter_date", nullable = false)
    private LocalDate encounterDate;


    // take note SOAP: Subjective, Objective, Assessment, Plan
    @Column(columnDefinition = "TEXT")
    private String subjective;

    @Column(columnDefinition = "TEXT")
    private String objective;

    @Column(columnDefinition = "TEXT")
    private String assessment;

    @Column(columnDefinition = "TEXT")
    private String plan;


    //vital signs
    private String bloodPressure;
    private Integer heartRate;
    private String temperature;
    private Integer spo2;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    @Builder.Default
    private EncounterStatus status = EncounterStatus.IN_PROGRESS;

    @OneToMany(mappedBy = "encounter", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<EncounterDiagnosis> diagnoses = new ArrayList<>();

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    public void addDiagnosis(EncounterDiagnosis diagnosis) {
        diagnoses.add(diagnosis);
        diagnosis.setEncounter(this);
    }
}